File: Circular_Linked_list/CircularDoublyLinkedListTraversal.java
// File Name: CircularDoublyLinkedListTraversal.java
// Program to traverse a circular doubly linked list
// forward and backward

class Node {
    int data;
    Node prev;
    Node next;

    Node(int data) {
        this.data = data;
    }
}

public class CircularDoublyLinkedListTraversal {
    static Node head = null;

    // Forward traversal
    static void traverseForward() {
        if (head == null) {
            System.out.println("List is empty");
            return;
        }

        Node temp = head;
        System.out.print("Forward traversal: ");
        do {
            System.out.print(temp.data + " <-> ");
            temp = temp.next;
        } while (temp != head);
        System.out.println("HEAD");
    }

    // Backward traversal
    static void traverseBackward() {
        if (head == null) {
            System.out.println("List is empty");
            return;
        }

        Node temp = head.prev;  // start from last node
        System.out.print("Backward traversal: ");
        do {
            System.out.print(temp.data + " <-> ");
            temp = temp.prev;
        } while (temp != head.prev);
        System.out.println("HEAD");
    }

    public static void main(String[] args) {
        traverseForward();
        traverseBackward();
    }
}

File: Circular_Linked_list/circular_doubly_linked_list_traversal.py
# File Name: circular_doubly_linked_list_traversal.py
# Program to traverse a circular doubly linked list
# forward and backward

class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

head = None

# Forward traversal
def traverse_forward():
    global head
    if head is None:
        print("List is empty")
        return

    temp = head
    print("Forward traversal:", end=" ")
    while True:
        print(temp.data, end=" <-> ")
        temp = temp.next
        if temp == head:
            break
    print("HEAD")

# Backward traversal
def traverse_backward():
    global head
    if head is None:
        print("List is empty")
        return

    temp = head.prev  # start from last node
    print("Backward traversal:", end=" ")
    while True:
        print(temp.data, end=" <-> ")
        temp = temp.prev
        if temp == head.prev:
            break
    print("HEAD")

if __name__ == "__main__":
    traverse_forward()
    traverse_backward()
